Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -349,6 +354,195 @@ private Task Upgrade2To3(Context context)
+            {
+                var result = await trackingContext.Context.PerformOperationAsync(
+                    Tracer,
+                    () =>
+                    {
+                        if (SelfCheckUpToDate())
+                        {
+                            return Task.FromResult(Result.Success(SelfCheckResult.UpToDate()));
+                        }
+
+                        return SelfCheckCoreAsync(trackingContext.Context);
+                    });
+
+                if (result)
+                {
+                    MarkSelfChecked(success: true);
+                }
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// A result of <see cref="SelfCheckAsync"/>.
+        /// </summary>
+        public struct SelfCheckResult
+        {
+            /// <summary>
+            /// Total number of invalid files.
+            /// </summary>
+            public int InvalidFiles { get; }
+
+            /// <summary>
+            /// Total number of files scanned on disk.
+            /// </summary>
+            public int TotalFilesScanned { get; }
+
+            /// <nodoc />
+            public SelfCheckResult(int invalidHashes, int totalFilesScanned) =>
+                (InvalidFiles, TotalFilesScanned) = (invalidHashes, totalFilesScanned);
+
+            /// <nodoc />
+            public static SelfCheckResult UpToDate() => new SelfCheckResult(invalidHashes: 0, totalFilesScanned: -1);
+
+            /// <inheritdoc />
+            public override string ToString() => $"InvalidHashes={InvalidFiles}, TotalFilesScanned={TotalFilesScanned}";
+        }
+
+        private async Task<Result<SelfCheckResult>> SelfCheckCoreAsync(OperationContext context)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            var contentHashes = ReadContentHashesFromDisk(context);
+            _tracer.Debug(context, $"SelfCheck: Enumerated {contentHashes.Count} entries by {stopwatch.ElapsedMilliseconds}ms.");
+
+            stopwatch.Restart();
+
+            int invalidEntries = 0;
+
+            foreach (var (contentHash, fileInfo) in contentHashes)
+            {
+                if (context.Token.IsCancellationRequested)
+                {
+                    context.TracingContext.Debug("Exiting self check because cancellation was requested.");
+                    break;
+                }
+
+                var (isValid, error) = await ValidateFileAsync(context, contentHash, fileInfo);
+                if (!isValid)
+                {
+                    _tracer.Warning(context, $"Found invalid entry in cache. Hash={contentHash}. {error}. Evicting the file...");
+                    await RemoveInvalidContentAsync(context, contentHash);
+                    invalidEntries++;
+                }
+            }
+
+            return Result.Success(new SelfCheckResult(invalidHashes: invalidEntries, totalFilesScanned: contentHashes.Count));

```

Comment:
contentHashes.Count is incorrect if we break from cancelation. Better to keep running total in local variable.